File: posixlib/src/main/resources/scala-native/linux/epoll.c
#if defined(SCALANATIVE_COMPILE_ALWAYS) || defined(__SCALANATIVE_POSIX_EPOLL)

#ifdef __linux__

#include <stddef.h>
#include <sys/epoll.h>

int scalanative_epoll_cloexec() { return EPOLL_CLOEXEC; }

int scalanative_epoll_ctl_add() { return EPOLL_CTL_ADD; }
int scalanative_epoll_ctl_mod() { return EPOLL_CTL_MOD; }
int scalanative_epoll_ctl_del() { return EPOLL_CTL_DEL; }

int scalanative_epollin() { return EPOLLIN; }
int scalanative_epollout() { return EPOLLOUT; }
int scalanative_epollrdhup() { return EPOLLRDHUP; }
int scalanative_epollpri() { return EPOLLPRI; }
int scalanative_epollerr() { return EPOLLERR; }
int scalanative_epollhup() { return EPOLLHUP; }

int scalanative_epollet() { return EPOLLET; }
int scalanative_epolloneshot() { return EPOLLONESHOT; }

// linux 3.5
// int scalanative_epollwakeup() { return EPOLLWAKEUP; }
// linux 4.5
// int scalanative_epollexclusive() { return EPOLLEXCLUSIVE; }

size_t scalanative_epoll_event_size() { return sizeof(struct epoll_event); }

void scalanative_epoll_event_set(struct epoll_event *ev, int idx,
                                 uint32_t events, uint64_t data64) {
    struct epoll_event *evidx = ev + idx;
    evidx->events = events;
    evidx->data.u64 = data64;
}

void scalanative_epoll_event_get(struct epoll_event *ev, int idx,
                                 uint32_t *events, uint64_t *data64) {
    struct epoll_event *evidx = ev + idx;
    if (NULL != events)
        *events = evidx->events;
    if (NULL != data64)
        *data64 = evidx->data.u64;
}

#endif // __linux__

#endif // __SCALANATIVE_POSIX_EPOLL

File: posixlib/src/main/scala/scala/scalanative/linux/epoll.scala
package scala.scalanative
package linux

/*
 * https://man7.org/linux/man-pages/man2/epoll_wait.2.html
 * https://man7.org/linux/man-pages/man2/epoll_create1.2.html
 */

import posix._
import unsafe._

@extern
@define("__SCALANATIVE_POSIX_EPOLL")
object epoll {

  import stdint._

  @name("scalanative_epoll_cloexec")
  def EPOLL_CLOEXEC: CInt = extern

  @name("scalanative_epoll_ctl_add")
  def EPOLL_CTL_ADD: CInt = extern
  @name("scalanative_epoll_ctl_mod")
  def EPOLL_CTL_MOD: CInt = extern
  @name("scalanative_epoll_ctl_del")
  def EPOLL_CTL_DEL: CInt = extern

  @name("scalanative_epollin")
  def EPOLLIN: CInt = extern
  @name("scalanative_epollout")
  def EPOLLOUT: CInt = extern
  @name("scalanative_epollrdhup")
  def EPOLLRDHUP: CInt = extern
  @name("scalanative_epollpri")
  def EPOLLPRI: CInt = extern
  @name("scalanative_epollerr")
  def EPOLLERR: CInt = extern
  @name("scalanative_epollhup")
  def EPOLLHUP: CInt = extern

  @name("scalanative_epollet")
  def EPOLLET: CInt = extern
  @name("scalanative_epolloneshot")
  def EPOLLONESHOT: CInt = extern

  def epoll_create(size: CInt): CInt = extern
  def epoll_create1(flags: CInt): CInt = extern

  def epoll_ctl(
      epfd: CInt,
      op: CInt,
      fd: CInt,
      event: CVoidPtr
  ): CInt = extern

  @blocking
  def epoll_wait(
      epfd: CInt,
      events: CVoidPtr,
      maxevents: CInt,
      timeoutMillis: CInt
  ): CInt = extern

  def scalanative_epoll_event_size(): CSize = extern

  def scalanative_epoll_event_set(
      ev: CVoidPtr,
      idx: CInt,
      events: uint32_t,
      data: uint64_t
  ): Unit = extern

  def scalanative_epoll_event_get(
      ev: CVoidPtr,
      idx: CInt,
      events: Ptr[uint32_t],
      data: Ptr[uint64_t]
  ): Unit = extern

}
